File: ctci/ctci-coin-change.pl
#!/usr/bin/env perl
use strict;

# Hacker Rank ctci-coin-change
#
# mitch@mitchjacksontech.com
# https://www.hackerrank.com/challenges/ctci-coin-change
#
# https://www.youtube.com/watch?v=jaNZ83Q3QGc
# Thanks for helping me understand a useful approach

my @in;push @in,$_ for <>;chomp for @in;
my ($change_amt,undef) = split / /,shift @in;
my @coins = sort{$a <=> $b} split / /,shift @in;
my @count = (1);

#print "Coins: ".join (' ',@coins)."\n";
#print "Amt: $change_amt\n";

for my $c (@coins) {
    for my $i (1..$change_amt) {
        #print "\$c:$c \$i:$i ";
        if ($i >= $c) {
            $count[$i] += $count[$i-$c];
        }
        #print join(" ",@count)."\n";
    }
}

print $count[$change_amt] || 0;
print "\n";

File: ctci/ctci-coin-change.py
#!/usr/bin/env python3

# Hacker Rank ctci-coin-change
#
# mitch@mitchjacksontech.com
# https://www.hackerrank.com/challenges/ctci-coin-change
#
# https://www.youtube.com/watch?v=jaNZ83Q3QGc
# Thanks for helping me understand a useful approach

import sys

def make_change(coins, n):
    counts = [0] * (n+1);
    counts[0] = 1;

    for c in coins:
        for i in range(1,n+1):
            if i >= c:
                counts[i] += counts[i-c]

    return counts[n]

n,m = input().strip().split(' ')
n,m = [int(n),int(m)]
coins = [int(coins_temp) for coins_temp in input().strip().split(' ')]
print(make_change(coins, n))
